### DISCUSSION TITLE
When magnified 200 times, the circle is not smooth

### DISCUSSION THREAD

**Discussion by pcbren1987** [pixijs/pixijs#8625] [RFC]
Does PIXIJS have a way to refresh the update to make the circle smoother
![image](https://user-images.githubusercontent.com/89827437/188134321-5247185d-dee6-48aa-b995-d92d2f8d7db0.png)

**Comment by ivanpopelyshev** [comment]
Try https://github.com/pixijs/graphics-smooth

**Reply by bigtimebuddy** [reply]
When Pixi draws a circle (or any curve) it actually converts it to vertices and straight lines. Pixi also takes into account the relative size of the curves. So drawing a larger curve will add more vertices, or drawing a smaller curve will add fewer.

If you're going to be doing extreme scaling of your curves, it's better to draw it at the large-ish size and then scale down. 

You can also customize or disable if the curves are adaptive or not.
[PIXI.GRAPHICS_CURVES](http://pixijs.download/dev/docs/PIXI.html#GRAPHICS_CURVES).

**Reply by pcbren1987** [reply]
We use PixiJS to draw PCB vector map, and the original image effect is still maintained when it needs to be zooms in and out, and the scaling ratio can be up to 1000 times. If the arc is composed of N small line segments, Very inefficient(Is there a more efficient solution specifically for circles). I wonder if there is any other solution
As shown in the figure below
![2323](https://user-images.githubusercontent.com/89827437/188155308-aceebd54-1b88-4d35-b73a-73c46b7dd3a9.gif)

**Reply by domis86** [reply]
@pcbren1987 regarding performace my first thought is: culling. You should not draw any objects that are outside of screen. So if you zoom in and you see only 1 element, then you should not draw any other elements. I suspect such optimization should give good impact.

**Reply by domis86** [reply]
Also: when user zooms out you should try to draw objects using less details (because he will not see any difference anyway).

so to summarize:
- on zoom out: draw many objects with less details
- on zoom in: draw only visible objects but with more details

**Reply by bigtimebuddy** [reply]
If your schematics are largely static, why not just use SVG and the DOM? Performance is not going to be great, but if you need 1000x scaling, that's going to give you the best visual at the expense of CPU. There are clever things to rendering at various resolutions in Pixi, to add culling, etc, but I would not say that your use-case is not common which is why we don't have a generalized solution for infinitely scaling graphics.

**Reply by pcbren1987** [reply]
Elements with the number of elements reaching more than 500K, including a large number of circles, are slower to use SVG, and will be blurred when the number is large

**Reply by pcbren1987** [reply]
Pixi circle drawing is to generate hundreds of small line segments from arcs and draw arcs. In order to achieve the effect of arcs, the data volume of geometry and graphicsdata is very large. When drawing a line, cap is set to pixi.line_ Cap.round, render according to square lines, and then use glsl to generate circles, which can improve efficiency
Whether this scheme is feasible and greatly reduces the amount of data？
![image](https://user-images.githubusercontent.com/89827437/188812848-8c6db007-7dce-4546-90db-ad8205e0e5ab.png)

**Reply by pcbren1987** [reply]
Whether PIXI can provide such a rendering method in Pixi.Graphics, which can greatly improve the efficiency of rendering circles and the large amount of data is expected to support this scheme in the next version

**Reply by bigtimebuddy** [reply]
Are you reusing GraphicsGeometry between Graphics?

**Reply by pcbren1987** [reply]
Before using these two methods to test the speed, for the circle speed is not ideal
https://github.com/pixijs/pixijs/discussions/8358
PIXI graphics.DrawCircle can support one more way of doing this as shown in the example below. Cap is set to pixi.line_cap. round, cap is also rendered this way. It's going as fast as it can
https://www.pixiplayground.com/#/edit/5bWjCjL70izwdVOncfzoM

**Reply by pcbren1987** [reply]
> Try https://github.com/pixijs/graphics-smooth
Thank you very much. It is very fast to use this library, but there is a problem
The graph is scaled and the actual graph width remains the same

According to the middle mouse button scroll to zoom, use the PIXI. Graphics PIXI. Smooth. SmoothGraphics effect is not the same,
See links for examples
https://codepen.io/pcbren1987/pen/vYjZmgV

**Comment by ivanpopelyshev** [comment]
You want production-grade stuff that doesn't exist in other webgl renderers, its a serious job. Post a bounty to add primitive circles in a plugin.
